### PR TITLE
fix(next): watch before initial build

### PR DESCRIPTION
**Pull request by NathanColosimo**
## Summary

- start and await the workflow filesystem watcher before the initial development build
- remember whether a source event arrives during startup
- run one full convergence rebuild after startup when an event was buffered
- close the startup watcher if initialization fails before normal HMR handlers take ownership

## Testing

- pnpm --filter @workflow/next build
- pnpm --filter @workflow/next test -- --run src/watch-rebuild.test.ts

## Stack

1. #3333: rebuild convergence and unit coverage
2. This PR: close the startup observation gap
3. #3471: end-to-end HMR coverage

<sub>Stack created with <a href="https://github.com/github/gh-stack">GitHub Stacks CLI</a> • <a href="https://gh.io/stacks-feedback">Give Feedback 💬</a></sub>

### CODE REVIEW

File: packages/next/src/builder-eager.ts [hunk 1]
```diff
@@ -70,8 +72,87 @@ export async function getNextBuilderEager(
...
+        }
+      };
+      if (watcher) {
+        watcher.on('add', markInitialBuildChanged);
+        watcher.on('change', markInitialBuildChanged);
+        watcher.on('unlink', markInitialBuildChanged);
+        watcher.on('error', (error) => {
+          console.error('Workflow dev watcher error', error);
+        });
+        await closeWatcherOnError(once(watcher, 'ready'));
```
[the remark this answers is not in the document]
Comment (NathanColosimo) [reply packages/next/src/builder-eager.ts:217]: This is intentional. Before `ready`, the watcher is establishing the coverage invariant that makes the initial build safe. An error during that scan means we cannot know that edits made during startup were observed, so logging and continuing would reopen the lost-edit race this PR fixes. `closeWatcherOnError` closes the incomplete watcher and fails startup; errors after `ready` still use the existing log-and-continue handler.

[the remark this answers is not in the document]
Comment (NathanColosimo) [reply packages/next/src/builder-eager.ts:217]: I am keeping this behavior. This PR relies on the watcher completing its initial scan before the initial build so no startup edits are missed. If chokidar emits EMFILE/EACCES during that scan, the watcher is incomplete and continuing would silently violate the guarantee. The persistent error listener still reports errors after readiness; startup failure is intentional.

File: packages/next/src/builder-eager.ts [hunk 2]
```diff
@@ -431,42 +524,39 @@ export async function getNextBuilderEager(
...
-              }
-              const extension = extname(normalizedPath);
-              if (extension && !isSourceFile(normalizedPath)) {
-                return true;
-              }
-              return hasIgnoredPathFragment(normalizedPath);
-            },
-          }
-        );
+        const startupWasStable = await closeWatcherOnError(runFullRebuild());
```
Comment (VaguelySerious) [inline packages/next/src/builder-eager.ts:534]: ### AI Review: Blocking

The startup rediscovery is unconditional, so every `next dev` start now pays for two full workflow builds instead of one.

Before this PR the startup path built once and then seeded the baseline with a single `sourceSnapshots = await snapshotSources()`. That line is gone, and `runFullRebuild()` is called here regardless of whether the watcher saw anything. `fullRebuild()` starts with `clearDiscoveredEntriesCache()`, so nothing from the first build is reused.

Measured on `workbench/nextjs-webpack` (153 workflows, 147 steps) with `WORKFLOW_DEV_HMR_LOGS=1` and a probe around `buildCombinedFunction`, on a run where the watcher emitted **zero** events:

```
WFPROBE buildCombinedFunction start 1789426457223
WFPROBE buildCombinedFunction end   1789426457760   (537ms)
workflow dev hmr: full rediscovery
WFPROBE buildCombinedFunction start 1789426458199
WFPROBE buildCombinedFunction end   1789426458490   (291ms)
workflow dev hmr: ready
workflow dev hmr: idle
```

The user-visible effect is a duplicated `✓ Compiled workflows in …` line and roughly 0.3-0.6s of extra startup on this app; it scales with workflow count.

The second build is only needed if something changed while the first one ran, and you already track that:

```ts
const startupWasStable =
  watchGeneration === 0
    ? ((sourceSnapshots = await snapshotSources()), true)
    : await closeWatcherOnError(runFullRebuild());
```

(or equivalent). That keeps the race fix and drops the cost when there is no race, which is the common case.

File: packages/next/src/builder-eager.ts [hunk 3]
```diff
@@ -123,8 +124,101 @@ export async function getNextBuilderEager(
...
+          ) {
+            throw new Error(`Unknown watch event: ${kind}`);
+          }
+          if (
+            mode === 'source' &&
+            (kind === 'addDir' || kind === 'unlinkDir')
+          ) {
+            return;
+          }
+          watchGeneration++;
```
Comment (VaguelySerious) [inline packages/next/src/builder-eager.ts:178]: ### AI Review: Blocking

`watchGeneration++` runs before anything checks whether the path is watchable, so unrelated files can invalidate a startup that was in fact stable.

The dependency watcher created in `replaceDependencyWatcher` passes `depth: 0` but **no `ignored` option**, so it reports every file in a dependency directory, not just the relevant ones. `isWatchableFile` does filter those out, but only later, inside `handleFileChanged` / `scheduleFullRebuild`. By then the generation has already moved.

That matters specifically during the startup window: `runFullRebuild()` captures `generation` on entry and compares `generation === watchGeneration` on exit. Any unwatchable write inside a watched dependency directory during that window flips `startupWasStable` to false and fires another `scheduleRebuild({ kind: 'full' })`.

The events are not hypothetical. Instrumenting the listener on `workbench/nextjs-webpack` captured Next's own generated files arriving through the dependency watcher:

```
workflow dev hmr: ready
workflow dev hmr: idle
✓ Generated AGENTS.md and CLAUDE.md for AI agents. …
WFPROBE event dependencies add …/workbench/nextjs-webpack/AGENTS.md
WFPROBE event dependencies add …/workbench/nextjs-webpack/CLAUDE.md
```

A control run appending to an unrelated `PROBE_NOTE.md` in the app root produced no rediscovery, which confirms the source watcher's `ignored` predicate is doing its job and the dependency watcher is the hole.

Separately, one uninstrumented run of the same app reached **four** `buildCombinedFunction` calls and three `full rediscovery` markers before settling, with no user edit. I did not instrument that particular run, so I can't attribute those extra rediscoveries to this path with certainty, and I'm not claiming it as proof. The mechanism above is established from the code and from the captured events regardless.

Two cheap fixes, and I'd take both:

1. Move the watchability filter ahead of the counter, so unwatchable paths never bump the generation.
2. Give the dependency watcher an `ignored` predicate (`(p) => !relevantFiles.has(normalizePath(String(p)))`, or reuse `isIgnoredWatchPath`) so it stops registering interest in the whole directory.

This also has a downstream effect on the test PR stacked on top of this one, which asserts an exact zero full-rediscovery count for hot-rebuild cases.

File: packages/next/src/builder-eager.ts [hunk 4]
```diff
@@ -328,38 +443,17 @@ export async function getNextBuilderEager(
...
-        };
-
         const runFullRebuild = async () => {
           const generation = watchGeneration;
           logDevHmr('workflow dev hmr: full rediscovery');
           const buildWasStable = await fullRebuild();
-          await refreshKnownFiles();
-          return buildWasStable && generation === watchGeneration;
+          await replaceDependencyWatcher();
+          const finalSnapshots = await snapshotSources();
```
Comment (VaguelySerious) [inline packages/next/src/builder-eager.ts:451]: AI Review: Note

Every full rebuild now stats the source tree three times: once at the top of `fullRebuild` (`nextSourceSnapshots`), once at its return to compare, and a third time here.

The second and third reads are separated only by `replaceDependencyWatcher()`, and they compute nearly the same predicate: `fullRebuild` compares `nextSourceSnapshots` against a fresh read, then this compares `sourceSnapshots` (just assigned `nextSourceSnapshots`) against another fresh read. On an app with a few hundred entries that is a redundant stat pass per rebuild, and rebuilds happen on every `add`/`unlink`.

If the intent is to also catch writes that land during `replaceDependencyWatcher()`, say so in a comment, because as written it reads like an accident. Otherwise reuse `buildWasStable`'s snapshot.

File: packages/next/src/builder-eager.ts [hunk 5]
```diff
@@ -123,8 +124,101 @@ export async function getNextBuilderEager(
...
+        (isAbsolute(pathname)
+          ? pathname
+          : resolve(this.config.workingDir, pathname)
+        ).replace(/\\/g, '/');
+      const normalizedWorkingDir = normalizePath(this.config.workingDir);
+      let relevantFiles = new Set<string>();
+      const isWatchableFile = (path: string) =>
+        isSourceFile(path) || relevantFiles.has(path);
+      const normalizedDistDir = normalizePath(this.config.distDir);
+      const isIgnoredWatchPath = createWatchIgnorePredicate({
```
Comment (VaguelySerious) [inline packages/next/src/builder-eager.ts:137]: AI Review: Nit

`createWatchIgnorePredicate` moved out of the `if (this.config.watch)` block, so production builds now construct it too. It calls `loadGitignoreMatchers`, which walks up to the project root doing synchronous `readFileSync` on each `.gitignore`.

The cost is small (a handful of sync reads) and it is only build-time, so this isn't worth reworking the control flow for. But it is a behavior change outside dev mode that the PR doesn't mention, and it makes a production build's output depend on `.gitignore` parsing that previously never ran there. `sourceWatcherCovers` is the only non-watch consumer; if it is itself only reachable under `watch`, the predicate can stay inside the block.

**Review by VaguelySerious — COMMENTED** [review]
AI review: blocking issues found

File: packages/next/src/builder-eager.ts (hunk 2 above)
Comment (VaguelySerious) [reply packages/next/src/builder-eager.ts:534]: (AI) Re-verified against the current branch head (43bf5b3b, after the merge of the updated base). This still stands, unchanged.

`builder-eager.ts:231` runs `this.buildCombinedFunction(options)` for the initial build. `builder-eager.ts:534` then runs `runFullRebuild()` unconditionally, and `fullRebuild()` opens with `clearDiscoveredEntriesCache()` at line 402 before calling `buildCombinedFunction` again at line 410. Nothing from the first build survives the cache clear, so every `next dev` start pays two full workflow builds even when the watcher emitted zero events.

A note on what makes this harder to fix than it looks, since the author is no longer available to: the obvious fix (build once, then seed the baseline with `sourceSnapshots = await snapshotSources()`) is the thing this stack deliberately removed. Per the reply on #3333, `sourceSnapshots` must contain "only bytes read before the accepted build", because a file can change between esbuild reading it and a post-build snapshot reading it, which would swallow the queued invalidation and leave the bundle stale.

So the correct fix is to route the *initial* build through the same snapshot-then-build ordering `fullRebuild()` already implements, rather than building at 231 and re-building at 534. That preserves the coverage invariant this PR exists to establish and removes the duplicate build. It is a real change to the startup path, not a one-liner, and it needs the Local Dev lanes to validate.

File: packages/next/src/builder-eager.ts (hunk 3 above)
Comment (VaguelySerious) [reply packages/next/src/builder-eager.ts:178]: (AI) Re-verified against 43bf5b3b. Still stands.

`watchGeneration++` is at `builder-eager.ts:178`, inside `attachWatchEvents`' `'all'` handler, before any watchability test. The only filtering ahead of it is the `addDir`/`unlinkDir` skip for `mode === 'source'`.

The source watcher is largely protected by its own `ignored` predicate (`stats?.isFile() === true && !isSourceFile(normalizedPath)`). The dependency watcher is not: `replaceDependencyWatcher` at line 326 creates it with `{ depth: 0, ignoreInitial: true, followSymlinks: true }` and **no `ignored` option**, so every file written in a watched dependency directory bumps the generation. `runFullRebuild` captures `generation` on entry (line 447) and compares at line 455, so any such write during the startup window flips `startupWasStable` to false and fires a third full rebuild on top of the two above.

Smallest fix is to gate the increment on the same predicate the handlers already use, e.g. skip when `mode === 'dependencies' && !isWatchableFile(normalizePath(pathname))`, or give the dependency watcher an `ignored` predicate matching the source watcher's. Either is small and independent of the startup-path work in the other blocker.
